### PR TITLE
[CDAP-19016] Optimize deserialization 

### PR DESCRIPTION
**Pull request by tivv**
* Reusing introspection results for unsafe object creation 
 * Removing reflection for enum access

### CODE REVIEW

File: cdap-common/src/main/java/io/cdap/cdap/common/lang/InstantiatorFactory.java [hunk 1]
```diff
@@ -174,14 +181,28 @@ public T create() {
   }
 
   private <T> Instantiator<T> getByUnsafe(final TypeToken<T> type) {
+    //We do introspection once and then just use reflection to set field values
+    List<Map.Entry<Field, Object>> fieldsToInitialize = new ArrayList<>();
+    Reflections.visit(null, type.getType(), new FieldVisitor() {
+      @Override
```
Comment (fernst) [inline cdap-common/src/main/java/io/cdap/cdap/common/lang/InstantiatorFactory.java]: nit: moving this FieldVisitor implementation to a new class

**Review by fernst — APPROVED** [review]
LGTM, just a nit.

File: cdap-common/src/main/java/io/cdap/cdap/common/lang/InstantiatorFactory.java [hunk 2]
```diff
@@ -174,14 +181,28 @@ public T create() {
...
         try {
           Object instance = UNSAFE.allocateInstance(type.getRawType());
-          Reflections.visit(instance, type.getType(), new FieldInitializer());
+          for (Map.Entry<Field, Object> e : fieldsToInitialize) {
+            e.getKey().set(instance, e.getValue());
+          }
           return (T) instance;
-        } catch (InstantiationException e) {
+        } catch (InstantiationException | IllegalAccessException e) {
           throw Throwables.propagate(e);
```
Comment (sanjanasandeep) [inline cdap-common/src/main/java/io/cdap/cdap/common/lang/InstantiatorFactory.java:189]: is this single pipe intentional?

Comment (tivv) [reply cdap-common/src/main/java/io/cdap/cdap/common/lang/InstantiatorFactory.java:189]: That's multicatch syntax, see https://www.geeksforgeeks.org/multicatch-in-java/

Comment (fernst) [reply cdap-common/src/main/java/io/cdap/cdap/common/lang/InstantiatorFactory.java:189]: Yes, this allows us to manipulate multiple exception types using the same code https://docs.oracle.com/javase/8/docs/technotes/guides/language/catch-multiple.html#multiple

Comment (sanjanasandeep) [reply cdap-common/src/main/java/io/cdap/cdap/common/lang/InstantiatorFactory.java:189]: ah I see, thanks!

**Review by sanjanasandeep — APPROVED** [review]

File: cdap-common/src/main/java/io/cdap/cdap/common/lang/InstantiatorFactory.java (hunk 1 above)
Comment (tivv) [reply cdap-common/src/main/java/io/cdap/cdap/common/lang/InstantiatorFactory.java]: Done

**Pull request merged**